cf-key/cf-key-functions.c

```diff
@@ -117,23 +129,98 @@ int PrintDigest(const char* pubkey)
     return 0; /* OK exitcode */
 }
 
-int TrustKey(const char* pubkey)
+/**
+ * Split a "key" argument of the form "[[user@]address:]filename" into
+ * components (public) key file name, IP address, and (remote) user
+ * name.  Pointers to the corresponding segments of the #keyarg
+ * string will be written into the three output arguments #filename,
+ * #ipaddr, and #username. (Hence, the three output string have
+ * the same lifetime/scope as the #keyarg string.)
+ *
+ * The only required component is the file name.  If IP address is
+ * missing, NULL is written into the #ipaddr pointer.  If the
+ * username is missing, #username will point to the constant string
+ * "root".
+ *
+ * @NOTE the #keyarg argument is modified by this function!
+ */
+void ParseKeyArg(char *keyarg, char **filename, char **ipaddr, char **username)
 {
-    char *digeststr = GetPubkeyDigest(pubkey);
-    char outfilename[CF_BUFSIZE];
-    bool ok;
+    char *s;
 
-    if (NULL == digeststr)
+    /* set defaults */
+    *ipaddr = NULL;
+    *username = "root";
+
+    /* use rightmost colon so we can cope with IPv6 addresses */
+    s = strrchr(keyarg, ':');
+    if (s == NULL)
+    {
+        /* no colon, entire argument is a filename */
+        *filename = keyarg;
+        return;
+    }
+
+    *s = '\0';              /* split string */
+    *filename = s + 1;      /* filename starts at 1st character after ':' */
+
+    s = strchr(keyarg, '@');
+    if (s == NULL)
+    {
+        /* no username given, use default */
+        *ipaddr = keyarg;
+        return;
+    }
+
+    *s = '\0';
+    *ipaddr = s + 1;
+    *username = keyarg;
+
+    /* special case: if we got user@:/path/to/file
+       then reset ipaddr to NULL instead of empty string */
+    if (**ipaddr == '\0')
+    {
+        *ipaddr = NULL;
+    }
+
+    return;
+}
+
+/**
+ * Trust the given key.  If @c ipaddress is not @c NULL, then also
+ * update the "last seen" database.  The IP address is required for
+ * trusting a server key (on the client); it is -currently- optional
+ * for trusting a client key (on the server).
+ */
+int TrustKey(const char* filename, const char* ipaddress, const char* username)
+{
+    RSA* key;
+    char *digest;
+
+    key = LoadPublicKey(filename);
+    if (NULL == key)
+    {
         return 1; /* ERROR exitcode */
+    }
 
-    snprintf(outfilename, CF_BUFSIZE, "%s%cppkeys%croot-%s.pub",
-             GetWorkDir(), FILE_SEPARATOR, FILE_SEPARATOR, digeststr);
+    digest = GetPubkeyDigest(key);
+    if (NULL == digest)
+    {
+        return 1; /* ERROR exitcode */
+    }
 
-    free(digeststr);
+    if (ipaddress != NULL)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "Adding a CONNECT entry in lastseen db: IP '%s', key '%s'",
+            ipaddress, digest);
+        LastSaw1(ipaddress, digest, LAST_SEEN_ROLE_CONNECT);
+    }
 
-    ok = CopyRegularFileDisk(pubkey, outfilename);
+    SavePublicKey(username, digest, key);
```

Comment:
If the save operation fails (e.g., read-only filesystem), the function still exists with code 0 (== OK). Can you correct this?